packages/js/client-config-builder/src/BaseClientConfigBuilder.ts

```diff
@@ -0,0 +1,348 @@
+      this.addPackages(config.packages);
+    }
+
+    if (config.resolvers) {
+      this.addResolvers(config.resolvers);
+    }
+
+    return this;
+  }
+
+  addWrapper(uriWrapper: IUriWrapper<Uri | string>): ClientConfigBuilder {
+    const wrapperUri = Uri.from(uriWrapper.uri);
+
+    const existingRegistration = this.config.wrappers.find((x) =>
+      Uri.equals(x.uri, wrapperUri)
+    );
+
+    if (existingRegistration) {
+      existingRegistration.wrapper = uriWrapper.wrapper;
+    } else {
+      this.config.wrappers.push({
+        uri: wrapperUri,
+        wrapper: uriWrapper.wrapper,
+      });
+    }
+
+    return this;
+  }
+
+  addWrappers(uriWrappers: IUriWrapper<Uri | string>[]): ClientConfigBuilder {
+    for (const uriWrapper of uriWrappers) {
+      this.addWrapper(uriWrapper);
+    }
+
+    return this;
+  }
+
+  removeWrapper(uri: Uri | string): ClientConfigBuilder {
+    const wrapperUri = Uri.from(uri);
+
+    const idx = this.config.wrappers.findIndex((x) =>
+      Uri.equals(x.uri, wrapperUri)
+    );
+
+    if (idx > -1) {
+      this.config.wrappers.splice(idx, 1);
+    }
+
+    return this;
+  }
+
+  addPackage(uriPackage: IUriPackage<Uri | string>): ClientConfigBuilder {
+    const packageUri = Uri.from(uriPackage.uri);
+
+    const existingRegistration = this.config.packages.find((x) =>
+      Uri.equals(x.uri, packageUri)
+    );
+
+    if (existingRegistration) {
+      existingRegistration.package = uriPackage.package;
+    } else {
+      this.config.packages.push({
+        uri: packageUri,
+        package: uriPackage.package,
+      });
+    }
+
+    return this;
+  }
+
+  addPackages(uriPackages: IUriPackage<Uri | string>[]): ClientConfigBuilder {
+    for (const uriPackage of uriPackages) {
+      this.addPackage(uriPackage);
+    }
+
+    return this;
+  }
+
+  removePackage(uri: Uri | string): ClientConfigBuilder {
+    const packageUri = Uri.from(uri);
+
+    const idx = this.config.packages.findIndex((x) =>
+      Uri.equals(x.uri, packageUri)
+    );
+
+    if (idx > -1) {
+      this.config.packages.splice(idx, 1);
+    }
+
+    return this;
+  }
+
+  addEnv(uri: Uri | string, env: Record<string, unknown>): ClientConfigBuilder {
+    const envUri = Uri.from(uri);
+
+    const idx = this.config.envs.findIndex((x) => Uri.equals(x.uri, envUri));
+
+    if (idx > -1) {
+      this.config.envs[idx].env = {
+        ...this.config.envs[idx].env,
+        ...env,
+      };
+    } else {
+      this.config.envs.push({
+        uri: envUri,
+        env: env,
+      });
+    }
+
+    return this;
+  }
+
+  addEnvs(envs: Env<Uri | string>[]): ClientConfigBuilder {
+    for (const env of envs) {
+      this.addEnv(env.uri, env.env);
+    }
+
+    return this;
+  }
+
+  removeEnv(uri: Uri | string): ClientConfigBuilder {
+    const envUri = Uri.from(uri);
+
+    const idx = this.config.envs.findIndex((x) => Uri.equals(x.uri, envUri));
+
+    if (idx > -1) {
+      this.config.envs.splice(idx, 1);
+    }
```

Comment:
Haven't written this code, but I suspect it's a bit faster than using filter since the arrays are always unique.
I'm fine with either solution
cc @pileks

Reply:
In that case, having a `Set` would be the fastest. I belive the loading will also be easier